### ISSUE TITLE
OpenJDK example gets stuck and does not finish

### ISSUE THREAD

**Issue by ColossusChang** [gramineproject/examples#26]
When I tried to run the [OpenJDK example](https://github.com/gramineproject/examples/tree/master/openjdk) with `gramine-direct`, the process would get stuck and would not finish. The following are the logs in debug mode.
```
[::] debug: Gramine was built from commit: c56a3536b182021fa0655d9d55f3076ceced6fff
[::] debug: Host: Linux
[::] debug: LibOS xsave_enabled 1, xsave_size 0xa88(2696), xsave_features 0x2e7
[::] debug: Initial VMA region 0x7ffff3d57000-0x7ffff3ff8000 (LibOS) bookkeeped
[::] debug: Initial VMA region 0x7ffff3ff9000-0x7ffff7ff9000 (pal_init_pool) bookkeeped
[::] debug: Initial VMA region 0x7ffff7ffd000-0x7ffff7fff000 (vdso) bookkeeped
[::] debug: Initial VMA region 0x7ffff7ff9000-0x7ffff7ffd000 (vvar) bookkeeped
[::] debug: Initial VMA region 0x7ffff3ff8000-0x7ffff3ff9000 (pal_internal_me) bookkeeped
[::] debug: ASLR top address adjusted to 0x1f0f88b57000
[::] debug: host is not Linux-SGX, skipping /dev/attestation setup
[::] debug: Shim loaded at 0x7ffff3d57000, ready to initialize
[::] debug: mounting "file:." (chroot) under /
[::] debug: mounting "proc" (pseudo) under /proc
[::] debug: mounting "dev" (pseudo) under /dev
[::] debug: mounting "dev:tty" (chroot) under /dev/tty
[::] debug: Creating pipe: pipe.srv:c0c58f539a175c77f2947775c282f9ea3722675aa7af71fc985268a1e9bb1af8
[P1:T1:] error: Detected deprecated syntax: 'fs.mount'. Consider converting to the new array syntax: 'fs.mounts = [{ type = "chroot", uri = "...", path = "..." }]'.
[P1:T1:] debug: mounting "file:/usr/local/lib/x86_64-linux-gnu/gramine/runtime/glibc" (chroot) under /lib
[P1:T1:] debug: mounting "file:/usr" (chroot) under /usr
[P1:T1:] debug: mounting "file:/lib/x86_64-linux-gnu" (chroot) under /lib/x86_64-linux-gnu
[P1:T1:java] debug: Creating pipe: pipe.srv:31bdce43385c24f3c79bea913870890bcbc8454dab29f108c2c4d7fb6a8bf5c7
[P1:T1:java] debug: Allocating stack at 0 (size = 262144)
[P1:T1:java] debug: loading "file:/usr/lib/jvm/jdk1.8.0_192/bin/java"
[P1:T1:java] debug: find_interp: searching for interpreter: /lib/ld-linux-x86-64.so.2
[P1:T1:java] debug: loading "file:/usr/local/lib/x86_64-linux-gnu/gramine/runtime/glibc/ld-linux-x86-64.so.2"
[P1:T1:java] debug: Creating pipe: pipe.srv:1
[P1:T1:java] debug: Creating pipe: pipe.srv:8f05ebc24082473aae349742b937924405af3788da25383703d785dc9cca2147
[P1:T1:java] debug: Shim process initialized
[P1:shim] debug: IPC worker started
[P1:T1:java] debug: glibc register library /lib/libpthread.so.0 loaded at 0x1f0f886e0000
[P1:T1:java] debug: glibc register library /usr/lib/jvm/jdk1.8.0_192/bin/../lib/amd64/jli/libjli.so loaded at 0x1f0f884c8000
[P1:T1:java] debug: glibc register library /lib/libdl.so.2 loaded at 0x1f0f882c6000
[P1:T1:java] debug: glibc register library /lib/libc.so.6 loaded at 0x1f0f87ed5000
[P1:T1:java] debug: glibc register library /usr/lib/jvm/jdk1.8.0_192/jre/lib/amd64/server/libjvm.so loaded at 0x1f0f86eee000
[P1:T1:java] debug: glibc register library /lib/libm.so.6 loaded at 0x1f0f86c18000
[P1:T1:java] warning: Unsupported system call 435
[P1:T1:java] debug: Creating pipe: pipe.srv:03021a728396c0b488e71c2660e46d67f6dc87ee54bc6e4b568c2743f398ed49
[P1:T1:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T1:java] warning: Non-private futexes are not supported, assuming implicit FUTEX_PRIVATE_FLAG
[P1:T2:java] debug: ipc send to 1: IPC_MSG_PID_GETMETA(2, CRED)
[P1:T2:java] debug: Sending ipc message to 1
[P1:T2:java] debug: Waiting for a response to 1
[P1:shim] debug: IPC worker: received IPC message from 1: code=8 size=25 seq=1
[P1:shim] debug: ipc callback from 1: IPC_MSG_PID_GETMETA(2, CRED)
[P1:shim] debug: IPC send to 1: shim_ipc_pid_retmeta{8, ...}
[P1:shim] debug: Sending ipc message to 1
[P1:shim] debug: IPC worker: received IPC message from 1: code=0 size=37 seq=1
[P1:shim] debug: Got an IPC response from 1, seq: 1
[P1:T2:java] debug: Waiting finished: 0
[P1:T2:java] debug: glibc register library /lib/librt.so.1 loaded at 0x1f0f86915000
[P1:T2:java] debug: glibc register library /usr/lib/jvm/jdk1.8.0_192/jre/lib/amd64/libverify.so loaded at 0x1f0f86706000
[P1:T2:java] debug: glibc register library /usr/lib/jvm/jdk1.8.0_192/jre/lib/amd64/libjava.so loaded at 0x1f0f864d7000
[P1:T2:java] debug: glibc register library /usr/lib/jvm/jdk1.8.0_192/jre/lib/amd64/libzip.so loaded at 0x1f0f862b1000
[P1:T2:java] debug: memory fault at 0x00000000 (IP = 0x1f0f710002b4)
[P1:T2:java] debug: Created sigframe for sig: 11 at 0x1f0f86c15a50 (handler: 0x1f0f877f68c0, restorer: 0x1f0f87f10ae0)
[P1:T2:java] warning: Unsupported system call 435
[P1:T2:java] debug: Creating pipe: pipe.srv:c45f85f71797b2a501dfe215219dbaf9745e05ac3fc6fdfbb89e063d1c716578
[P1:T3:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Unsupported system call 435
[P1:T2:java] debug: Creating pipe: pipe.srv:5b1533b28b4591621d8598d8a33b4a69e0086a4bb04c77380c5d3db371dcf90c
[P1:T4:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T4:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Unsupported system call 435
[P1:T4:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] debug: Creating pipe: pipe.srv:6481f08d9a0e3adb5b0a628ed1d1823a8abe8c31386cbc45df9077e1ba8162f3
[P1:T5:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T5:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T5:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Unsupported system call 435
[P1:T2:java] debug: Creating pipe: pipe.srv:64fe1c9eca5f8e6f6352060336c0d1e777ae865b83bccd1085f85885e1cd0970
[P1:T6:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Unsupported system call 435
[P1:T2:java] debug: Creating pipe: pipe.srv:371d2ce26743f59154adc206e42d9a295047fc7f0130e96a75047bdd48115ec1
[P1:T6:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T7:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Unsupported system call 435
[P1:T2:java] debug: Creating pipe: pipe.srv:6c85a011368d947e2d71a3d8ee4068e5b60bbd691cf10ddba251f67b4d15bfcd
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T8:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T8:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T8:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T8:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Unsupported system call 435
[P1:T2:java] debug: Creating pipe: pipe.srv:ba0cd14bdce7d14b5ebc7aabf923c9eab886f23c9bfa83f86cd719fbcc8bfd01
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T9:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T9:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Unsupported system call 435
[P1:T2:java] debug: Creating pipe: pipe.srv:fec0095561d910c722b5dcd465ec08125830d93791b87c54176d062094acc418
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T10:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T7:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T8:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
[P1:T2:java] warning: Ignoring FUTEX_CLOCK_REALTIME flag
```

**Comment by ColossusChang** [comment]
Running `gramine-sgx` with `GRAMINE_LOG_LEVEL = trace` also ended up in being stuck. Here is the lo
[sgx.log](https://github.com/gramineproject/examples/files/8706677/sgx.log)
g

**Comment by dimakuv** [comment]
I just verified our OpenJDK example with both `gramine-direct` and `gramine-sgx`.

I didn't modify anything in the OpenJDK example. I used the latest Gramine version (commit https://github.com/gramineproject/gramine/commit/e6b697521a35ae24809d79237ad814ca497b3470).

I tested like this:
```sh
# first install OpenJDK and build
$ sudo apt-get install openjdk-11-jdk
$ SGX_SIGNER_KEY=<my-enclave-key> make SGX=1


# run native version
gramine-direct java MultiThreadMain
[P1:T1:] error: Detected deprecated syntax: 'fs.mount'. Consider converting to the new array syntax: 'fs.mounts = [{ type = "chroot", uri = "...", path = "..." }]'.
Final Count is: 10000


# run SGX version -- on my weak machine, it took *3 minutes* to finish
$ gramine-sgx java -Xmx8G MultiThreadMain
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

[P1:T1:] error: Detected deprecated syntax: 'fs.mount'. Consider converting to the new array syntax: 'fs.mounts = [{ type = "c
hroot", uri = "...", path = "..." }]'.
Final Count is: 10000
```

The test was done on Ubuntu 18.04. The installed OpenJDK version is: 
```
$ /usr/lib/jvm/java-11-openjdk-amd64/bin/java --version
openjdk 11.0.15 2022-04-19
OpenJDK Runtime Environment (build 11.0.15+10-Ubuntu-0ubuntu0.18.04.1)
OpenJDK 64-Bit Server VM (build 11.0.15+10-Ubuntu-0ubuntu0.18.04.1, mixed mode, sharing)
```

**Comment by dimakuv** [comment]
@ColossusChang So I suggest:

1. Please try with the latest Gramine commit.
2. Please tell us if you made any changes to the `java.manifest.template` file or to anything else.
3. Please show us the command lines that you use to build and run this example.
4. Please check the Java version you have installed.

**Comment by dimakuv** [comment]
Also note that on my pretty slow machine, the SGX version took 3 minutes to finish. (On the other side, `gramine-direct` version finished immediately, in 1 second.)

**Comment by llly** [comment]
I tried some commits and found that the hang issue of OpenJDK example did exist in gramine master for some time. It was fixed a few weeks ago by [#gramine/530](https://github.com/gramineproject/gramine/pull/530)

**Comment by dimakuv** [comment]
Thanks @llly for checking. Since this is fixed in the latest master, and will be available in the stable release of Gramine (v1.2) very soon, I'm closing this issue.
